include/RED4ext/Relocation-inl.hpp

```diff
@@ -27,50 +26,207 @@ uintptr_t RED4ext::UniversalRelocBase::Resolve(uint32_t aHash)
+                   L"\n"
+                   L"If you are the mod's developer, verify that your mod was loaded by RED4ext. "
+                   L"Alternatively, you may need to provide your own address resolver.";
+
+        ShowErrorAndTerminateProcess(msg, GetLastError());
+    }
+
+    return handle;
+}
+
+RED4EXT_INLINE RED4ext::UniversalRelocBase::ResolveFunc_t RED4ext::UniversalRelocBase::
+    InitializeAddressResolverFunction()
+{
+    constexpr auto procName = "RED4ext_ResolveAddress";
+
+    auto handle = GetRED4extModule();
+
+    auto func = reinterpret_cast<ResolveFunc_t>(GetProcAddress(handle, procName));
+    if (func == nullptr)
+    {
+        auto msg = L"The mod you are using is unable to find the required address resolver function from RED4ext.\n"
+                   L"This may occur if RED4ext is not properly loaded, OR if the mod is incompatible with the current "
+                   L"version of RED4ext.\n"
+                   L"\n"
+                   L"Please ensure that RED4ext is correctly installed AND that both RED4ext and the mod are "
+                   L"up-to-date.";
+
+        ShowErrorAndTerminateProcess(msg, GetLastError());
+    }
+
+    return func;
+}
+
+RED4EXT_INLINE RED4ext::UniversalRelocBase::ResolveFunc_t RED4ext::UniversalRelocBase::GetAddressResolverFunction()
+{
+    static const ResolveFunc_t func = InitializeAddressResolverFunction();
+    return func;
+}
+
+RED4EXT_INLINE HMODULE RED4ext::UniversalRelocBase::GetCurrentModuleHandle()
+{
+    HMODULE result;
+
+    if (!GetModuleHandleExW(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT,
+                            reinterpret_cast<LPCWSTR>(UniversalRelocBase::Resolve), &result))
+    {
+        auto msg =
+            L"Unable to retrieve the handle for a plugin.\n"
+            L"Normally, this issue should not happen.\n"
+            L"\n"
+            L"What you can do:\n"
+            L"    1. Disable all mods.\n"
+            L"    2. Enable them one by one.\n"
+            L"    3. Start the game after each change to see if the issue happens again.\n"
+            L"\n"
+            L"For more detailed instructions on identifying the mod causing the issue, visit the following link:\n"
+            L"    https://tinyurl.com/2zvuctb6\n"
+            L"\n"
+            L"By following these instructions, you can identify the mod causing the issue and report it to the mod "
+            L"author for further assistance.";
+
+        MessageBoxW(nullptr, msg, L"RED4ext.SDK", MB_ICONERROR | MB_OK);
+        TerminateProcess(GetCurrentProcess(), 1);
+    }
+
+    return result;
+}
+
+RED4EXT_INLINE std::filesystem::path RED4ext::UniversalRelocBase::GetCurrentModulePath()
+{
+    constexpr auto pathLength = MAX_PATH + 1;
+    auto handle = GetCurrentModuleHandle();
+
+    std::wstring fileName;
+    do
+    {
+        fileName.resize(fileName.size() + pathLength, L'\0');
```

Comment:
oh i see now what you did... welp, this looks even worse i suppose in the above context now but unsure how to tune it quickly to be more robust... just leaving this as a note for now.

Reply:
kinda stupid this doesnt return size automatically like bunch of other Win32 API functions do when you supply nullptr + 0 instead of buffer and size...

Reply:
> kinda stupid this doesnt return size automatically like bunch of other Win32 API functions do when you supply nullptr + 0 instead of buffer and size...

I know right, what is this stupid stuff, just gib me the path ffs, the string I can free it later if needed.